ocaml/xcp-rrdd/lib/plugin/utils.ml

```diff
@@ -45,19 +48,25 @@ let exec_cmd (module D : Debug.DEBUG) ~cmdstring ~(f : string -> 'a option) =
         (h, t)
   in
   let pid =
-    Forkhelpers.safe_close_and_exec None (Some out_writeme) None [] cmd args
+    Forkhelpers.safe_close_and_exec None (Some out_writeme) (Some err_writeme)
+      [] cmd args
   in
   Unix.close out_writeme ;
-  let in_channel = Unix.in_channel_of_descr out_readme in
-  let vals = ref [] in
-  let rec loop () =
-    let line = input_line in_channel in
-    let ret = f line in
-    (match ret with None -> () | Some v -> vals := v :: !vals) ;
-    loop ()
+  Unix.close err_writeme ;
+  let read_and_close f fd =
+    let in_channel = Unix.in_channel_of_descr fd in
+    let vals = ref [] in
+    let rec loop () =
+      let line = input_line in_channel in
+      let ret = f line in
+      (match ret with None -> () | Some v -> vals := v :: !vals) ;
+      loop ()
+    in
+    (try loop () with End_of_file -> ()) ;
+    Unix.close fd ; List.rev !vals
```

Comment:
Should this not be done with `finally`? We will not reach this if we have an exception different from `End_of_file`. I also believe we have a function to iterate over a file line by line - but maybe it's not available here.

* `lines_fold` in `ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml`